File: c/reverseList.cpp
// https://leetcode.com/problems/reverse-linked-list/

#include <iostream>

#include "ListNode.hpp"
using namespace std;

class Solution {
   public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }

        ListNode* next = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return next;
    }
};

int main() {
    Solution solution;
    ListNode* head = new ListNode(
        1, new ListNode(
               2, new ListNode(3, new ListNode(4, new ListNode(5, nullptr)))));
    printList(head);
    ListNode* result = solution.reverseList(head);
    printList(result);
    return 0;
}

File: rust/src/bin/reverse_list.rs
// https://leetcode.com/problems/reverse-linked-list/
// 2024/03/21

mod list_node;
use crate::list_node::*;

impl Solution {
    pub fn reverse_list(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut prev: Option<Box<ListNode>> = None;

        while let Some(mut current) = head {
            head = current.next.take();
            current.next = prev;
            prev = Some(current);
        }

        return prev;
    }
}

pub struct Solution {}

fn main() {
    println!(
        "{:?}",
        ListNode::to_array(Solution::reverse_list(ListNode::from_array(&[
            1, 2, 3, 4, 5
        ])))
    ); // [5,4,3,2,1]

    println!(
        "{:?}",
        ListNode::to_array(Solution::reverse_list(ListNode::from_array(&[1, 2])))
    ); // [2,1]

    println!(
        "{:?}",
        ListNode::to_array(Solution::reverse_list(ListNode::from_array(&[])))
    ); // []
}
